### DISCUSSION TITLE
pipe into alias

### DISCUSSION THREAD

**Discussion by fellnerse** [nushell/nushell#4225] [Q&A] [answered]
Hello!
I'm wondering how to pipe a string into the `alias` command to execute it. I want to load my `gitaliases`.
```
nuopen ./.gitconfig | lines |find = | str trim | str find-replace ' =.*' '' | each {echo $"g($it) = git ($it)"}
────┬────────────────
  0 │ gl = git l
  1 │ ga = git a
  2 │ gap = git ap
  3 │ gc = git c
  4 │ gca = git ca
  5 │ gcm = git cm
  6 │ gcam = git cam
  7 │ gm = git m
  8 │ gd = git d
  9 │ gds = git ds
 10 │ gdc = git dc
 11 │ gs = git s
 12 │ gco = git co
 13 │ gcob = git cob
 14 │ gb = git b
 15 │ gba = git ba
 16 │ gla = git la
────┴────────────────
```
So this seems to work and now I would like to alias all of those commands but replacing echo with alias does not work:
```
❯ nuopen ./.gitconfig | lines |find = | str trim | str find-replace ' =.*' '' | each {alias $"g($it) = git ($it)"}
error: Invalid alias syntax
    ┌─ shell:136:100
    │
136 │ nuopen ./.gitconfig | lines |find = | str trim | str find-replace ' =.*' '' | each {alias $"g($it) = git ($it)"}
    │                                                                                                    ^ Expected space on both sides of =

```
Do you have an idea how to make this right? do i have to pipe it differently? or is it not possible to use a variable for alias? I was not able to execute alias with an env variable either...

I think the problem lies here:
```
❯ ba=ba alias $nu.env.ba = git $nu.env.ba | alias | lines |find ba
$nu.env.ba = 'git $nu.env.ba'
```

Thanks in advance!

**Comment by kubouch** [comment, marked as answer]
Hi, 
you won't be able to dynamically create aliases like that. That is a limitation of the language.

What I would suggest is to write the `alias gl = git l` etc. lines into a .nu file and source that file on startup.

**Reply by fellnerse** [reply]
I was able to solve it like this: 

```
nuopen ~/.gitconfig | lines | find = | str trim | str find-replace ' =.*' '' | each { $'alias g($it) = git ($it)'} | str collect (char newline) | save '/tmp/nustuff.nu'
source '/tmp/nustuff.nu'
```

It's a bit weird that you can not have a dynamic name for the sourceing and that when you source this file the tmp file also need do exist otherwise it will fail (probably a sanity check in sourcing files). This means that you can not create the tmp file here either, but it needs to be created before sourcing the file containing above code.

**Reply by kubouch** [reply]
Unfortunately, sourcing dynamically created file names is another limitation of the language. We would have to significantly redesign the language in order to allow that which is not planned currently.

**Reply by fellnerse** [reply]
Ok interesting. I have experimented now a lot with nu and really like it, but sourcing feels really painful. Two things I'm missing: sourcing dynamic files as mentioned above and the possibility to change environment variables globally. I do not really understand why this is contextualised. I have the feeling this introduces more limitations as advantages. Maybe I'm missing something here.

**Reply by kubouch** [reply]
The reason for no dynamic sourcing is that we split the script execution to parsing and evaluation steps and these are strictly separate. Sourcing is done in the parsing step where you cannot use any value from the evaluation step. That kind of simplifies the interpreter and makes it possible to have a standalone parser that can be used, e.g., for syntax highlighting.

Nushell as a language is immutable and follows scoping rules while other shells commonly mutate the global state. While mutating global state makes things easier at first, it does not scale to larger scripts -- too fragile. Also, inn the new engine we have a `par-each` command that is like `each` but runs the iterations in parallel. If we allowed the global environment to be changed from anywhere, we would run into a lot of issues with that command. So it's more like a language design thing. If we allowed mixing parse and eval steps and modifying global environment, Nushell would be a completely different language.

Maybe can you describe what you were trying to do with the global environment variables?

Note that I'm not the main designer of the language but this is what I gathered so far :-).

**Reply by fellnerse** [reply]
Thank you for the deeper insight! I switched to discord in the help channel jt was able help me and understand that I should use load-env instead of sourcing; thats something I did not got out of the docs

**Reply by kubouch** [reply]
Yeah, that command was introduced relatively recently and nobody probably put it in the docs.
